src/afrolabs/components/kafka/utilities.clj

```diff
@@ -130,11 +145,21 @@
       (ig/halt! system)
       (info "System done shutting down.")
 
+      (when stream-ch
+        (csp/close! stream-ch))
+
       ;; return value
-      (or (when (and nr-msgs
-                     (number? nr-msgs))
-            (vec (take nr-msgs (apply concat @loaded-msgs))))
-          (apply concat @loaded-msgs))
+      (when collect-messages?
+        (let [all-msgs (persistent! @loaded-msgs)]
+          (or (when (and nr-msgs
+                         (number? nr-msgs))
```

Comment:
You don't need this `and`. `number?` doesn't throw if it encounters a `nil`

On another note, not being able to do the `or` inside the transducer annoys me. I feel it should be possible, I just can't see it.

Reply:
That is worse, I suspect the only way I could get it as clean as I want is a macro, which would be wrong for just this.